tests/complex_workflow_test.py

```diff
@@ -0,0 +1,105 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from dataclasses import dataclass
+from typing import Callable, List, NewType
+
+import dask
+import numpy as np
+
+import sciline as sl
+
+# We use dask with a single thread, to ensure that call counting below is correct.
+dask.config.set(scheduler='synchronous')
+
+
+@dataclass
+class RawData:
+    data: np.ndarray
+    monitor1: float
+    monitor2: float
+
+
+SampleRun = NewType('SampleRun', int)
+BackgroundRun = NewType('BackgroundRun', int)
+DetectorMask = NewType('DetectorMask', np.ndarray)
+DirectBeam = NewType('DirectBeam', np.ndarray)
+SolidAngle = NewType('SolidAngle', np.ndarray)
+Raw = sl.parametrized_domain_type('Raw', RawData)
+Masked = sl.parametrized_domain_type('Masked', np.ndarray)
+IncidentMonitor = sl.parametrized_domain_type('IncidentMonitor', float)
+TransmissionMonitor = sl.parametrized_domain_type('TransmissionMonitor', float)
+TransmissionFraction = sl.parametrized_domain_type('TransmissionFraction', float)
+IofQ = sl.parametrized_domain_type('IofQ', np.ndarray)
+BackgroundSubtractedIofQ = NewType('BackgroundSubtractedIofQ', np.ndarray)
+
+
+def reduction_factory(tp: type) -> List[Callable]:
+    def incident_monitor(x: Raw[tp]) -> IncidentMonitor[tp]:
+        return IncidentMonitor[tp](x.monitor1)
```

Comment:
It would be nice if we could avoid repeating the return type. The injector should be able to convert the return value to the correct type. (Would likely need modifications to injector package.) But would this mess up type hinting?

Reply:
At least for run-time, you can just omit it. I don't know if the type-checker needs it.

Reply:
Tried it and neither mypy nor pyre like it when I just return the base type. So I guess we need to explicitly wrap the result.